### ISSUE TITLE
Silk GUI

### ISSUE THREAD

**Issue by zodern** [NodeOS/NodeOS#79]
I have started making a GUI in javascript named Silk.  You can find the github repository [here](https://github.com/zodern/Silk).
#### 4 main goals:
- Fast and Lightweight
- Designed with developers in mind
- Experiment with new ideas for GUI's
- Client (window manager) can be accessed remotely
#### What is Completed

I  duplicated Meteor.methods since they make it very for the client and server to communicate. 

The basics of window management are done.  It is possible to open and minimize windows.  It applies the correct z-index to windows.  

It is possible to remotely access Silk by going to the url shown in the terminal.
#### Plans

The first version will have a very simple text editor and file explorer.

There is no security to prevent anyone from remotely accessing the computer, which needs to be fixed.  We could look at [logon](https://github.com/piranna/logon).

Right now the window manager uses iframes to show the apps.  Some websites don't allow themselves to display in iframes, including [google.com](http://google.com), [jw.org](http://jw.org) and [github.com](http://github.com).  If the expo browser is used in node os, maybe the window manager could create exoframes (I think that is what they are called) for each app.

To try it, past these commands into your terminal.  

`npm i -g silk-gui` 
`cd silk`
`npm start`
#### Loading Screen

![screen shot 2014-11-04 at 11 47 39 am](https://cloud.githubusercontent.com/assets/8175744/4905158/58d20c0e-644d-11e4-8f5d-e2805577026c.png)
#### Desktop

![screen shot 2014-11-04 at 11 46 36 am](https://cloud.githubusercontent.com/assets/8175744/4905159/58d574f2-644d-11e4-923d-85ba136a55cc.png)
#### Taskbar Hover

Notice that the taskbar background becomes solid on hover.

![screen shot 2014-11-04 at 11 47 55 am](https://cloud.githubusercontent.com/assets/8175744/4905160/58d64440-644d-11e4-8e4c-5a3368e82817.png)

![screen shot 2014-11-04 at 11 48 30 am](https://cloud.githubusercontent.com/assets/8175744/4905157/58d22da6-644d-11e4-93a9-bc72b3cab34d.png)
#### Current State of File Explorer

![screen shot 2014-11-04 at 11 48 16 am](https://cloud.githubusercontent.com/assets/8175744/4905161/58d78602-644d-11e4-8697-f17127ae9e4e.png)

**Comment by serapath** [comment]
Maybe it's a stupid question, but would it be possible to merge or blend the shell/terminal with the desktop UI?
several approaches exist:
* https://github.com/substack/bashful
* https://github.com/substack/exterminate
* https://github.com/substack/browser-terminal
* https://github.com/petethepig/devtools-terminal
* http://acko.net/blog/on-termkit/
* https://github.com/unconed/TermKit + https://github.com/FireyFly/TermKit
* https://github.com/shepik/wkterm
* https://github.com/featurist/shellsubstitute
* https://github.com/arturadib/shelljs
* http://www.htmlfivewow.com/demos/terminal/terminal.html
* http://www.masswerk.at/jsuix/
* http://www.masswerk.at/termlib/
* https://code.google.com/p/shellinabox/
* https://github.com/mitotic/graphterm
* https://github.com/hoeck/schirm
* https://github.com/liftoff/GateOne + http://liftoffsoftware.com/ + http://liftoffsoftware.com/Products/GateOne
* http://wterm.saverpigeeks.com/

**Comment by piranna** [comment]
@serapath, I don't see the point of your links, the only interesting one I've found is ```wkterm```. Could you be able to explain what are you trying to achieve?

**Comment by serapath** [comment]
It was just meant for inspiration.
The NodeOS project excites me, but my technical knowledge is to shallow to contribute something more useful. I was playing around with some of those projects i linked to above and like them.
So in case these projects were unknown here - I was hoping to contribute some inspiration at least.
I'll keep quiet now - sorry for bothering.

**Comment by piranna** [comment]
Don't worry, you didn't bothered, only I didn't get the point of them :-) NodeOS is mostly a server OS, and we try to promote that GUI apps being build as clients that connect to a headless NodeOS instance using standard protocols to increase portability. The links that you posted are OK, but they can work on their own. They don't have anything to do with NodeOS, but definitely they can work with it :-)

**Comment by formula1** [comment]
@serapath If you want to help out with terminal or a UI, check out [Silk](https://github.com/zodern/Silk). Theres always more to do there and not all of it is particularly technical. 

However, when it comes to actually creating a gui, almost everything is based off a chrome/browser implementation and its far easier to build a terminal interface and make sure that the operating system internals are working fine then integrating a chrome shell. @netlovers has made a certian amount of progress, however at the end of the day he's doing a very difficult job. On top of that, he's doing it for free which means he should be under no obligation and only be celebrated for what success he does gain.

That being said, having a workable terminal, filesystem, process system and etc is truly what a javascript OS is, the gui is extra on top that will make people want to use it more.

**Comment by netlovers** [comment]
I made some more progress since then. As it turns out I can't use the Ozone-GBM platform inside a VMware vm, because the graphic driver there does not support it. I'm not sure if any vm supports GBM at the moment. But Dri uses software renderer so it might be enough for some simple stuff. Although my last attempt to build a Chrome version with the Dri platform failed.
I'm far away from my laptop for a few more days, so no progress will be made by me until I get back home.

I'm thinking about using real hardware instead of just a vm, because it looks like that a vm has some serious limitations, and I can't be sure if something won't work because it runs inside the vm, or if I'm facing a real bug. And to be honest, compiling Chrome inside a vm is just too painfully slow, but I'm not ready to buy a new build server just for this task.

**Comment by lawsonry** [comment]
I've been working on a NodeOS Shell project that bypasses the need to use a browser in order to render a terminal. My goal is to have a command line similar to our traditional terminals but inside you can run a command like `edit soandso.js` and it will call `edit.js` in some `/usr/share` folder and pass along the arguments to it. 

Ultimately, I'm trying to do two things: 

1. Create a nano clone for NodeOS that will be a simple CLI text editor. 

2. Create a vim clone for NodeOS so I can work exclusively inside NodeOS. 

3. Package a nano clone, a vim clone, and Slap together as the trifecta of NodeOS CLI editors. 

All of this requires the ability to produce a terminal-like system, which itself requires us to be able to manipulate the console which requires (I think) more integration on NodeOS's part with the actual mo'chine. 

On Linux, you can enter a Node environment by typing `node` and hitting enter. Boom. Now you're basically inside an interactive console application. But we don't need a JavaScript console; we need something like `Nash` (just made that up), which is a Bash/terminal specifically for NodeOS. I think this is the key to a proper CLI interface for NodeOS. 

`</ramble>`

**Comment by netlovers** [comment]
@lawsonry Check this out https://github.com/groundwater/node-bin-nsh

**Comment by piranna** [comment]
> I'm thinking about using real hardware instead of just a vm, because it looks like that a vm has some serious limitations, and I can't be sure if something won't work because it runs inside the vm, or if I'm facing a real bug. And to be honest, compiling Chrome inside a vm is just too painfully slow, but I'm not ready to buy a new build server just for this task.

What about a chroot system, like debootstrap? This way you have a
somewhat "pristine minimal" system that don't depend on your system
installed libraries but run on top of your native hardware... This way
you could compile them faster, and theorically it could also allow you
to test them... :-)


-- 
"Si quieres viajar alrededor del mundo y ser invitado a hablar en un
monton de sitios diferentes, simplemente escribe un sistema operativo
Unix."
– Linus Tordvals, creador del sistema operativo Linux

**Comment by piranna** [comment]
@lawsonry, do you have your shell code published somewhere?

**Comment by netlovers** [comment]
@piranna I'm using a Windows laptop with uefi, and fastboot, so normal linux distros won't really work out of box (https://help.ubuntu.com/community/UEFI). And I hate the current Linux desktops, so installing one is not an easy decision to make. :D

**Comment by piranna** [comment]
> @piranna I'm using a Windows laptop with uefi, and fastboot, so normal linux distros won't really work out of box (https://help.ubuntu.com/community/UEFI).

Oh, though you were using Linux already...


> And I hate the current Linux desktops, so installing one is not an easy decision to make. :D

Ubuntu Mate is your friend ;-)
https://ubuntu-mate.org/


-- 
"Si quieres viajar alrededor del mundo y ser invitado a hablar en un
monton de sitios diferentes, simplemente escribe un sistema operativo
Unix."
– Linus Tordvals, creador del sistema operativo Linux

**Comment by piranna** [comment]
Ok, so the emulated Cirrus of QEmus [has DRM support](http://cateee.net/lkddb/web-lkddb/DRM_CIRRUS_QEMU.html)... :-P

**Comment by netlovers** [comment]
It's okay, but I can't really test it, because Qemu on Windows is slow as h*ll.

**Comment by piranna** [comment]
Well, enabling ```CONFIG_FB_CIRRUS``` to have framebuffer support gets it to be slower that disabling KVM and didn't get the Unicode characters to work, only to show the penguin at boot... :-P

(Obviously, I've reverted it...)

**Comment by piranna** [comment]
@netlovers, I have just found [the compilation flags of Mesa EGL](http://www.mesa3d.org/egl.html). Probably you have seen earlier, but I've seen that it supports fbdev as EGL platform, and seems we have it by free (yes, ```CONFIG_FB_CIRRUS``` is slow as h*ll, but default one using vesafb runs awesome in QEmu :-D ) maybe it would be used to start moving one step at a time :-) If we can be able to run EGL at least software rendered we could be able to exec some of the ```node-egl``` demos I put earlier in this thread and later try to compile OpenVG and adapt node-canvas-openvg to have 2D and later 3D graphics on the console framebuffer, and later try to use DRI. This would be a longer path to have the things working, but probably it will be easier and also all this steps are minor milestones in the path to get CEF to get running, what do you think? :-)

**Comment by netlovers** [comment]
Why fbdev is necessary? I think Mesa should work fine with the flags that I used in my previous gist. The problem I have is with Chrome. I even thought about using X until the DRI platform is stable enough, but I know you are against X.

**Comment by piranna** [comment]
> Why fbdev is necessary? I think Mesa should work fine with the flags that I used in my previous gist. The problem I have is with Chrome.

Oh sorry, I though you were having problems using DRI with Mesa + EGL, that's why I proposed to use fbdev instead as a short term solution, but if you have got advances on this field then fbdev is out of place in benefict of DRI :-D Don't know if we could get big text resolutions without using fbdev (I need to test to use pure text-mode resolutions on the boot params) nor if it would be required to initialize graphic card to draw using DRI + EGL, but if not, then framebuffer could be fully removed.

Do you have somewhere your current buid scripts? This way I could check out how feasable would be to make [node-openvg-canvas](https://github.com/luismreis/node-openvg-canvas) to work that could be a good sandbox until CEF is available ([SVG fonts](https://github.com/SinisterMinister/node-openvg-canvas/commit/f52f33c4d988944ed45cb5331fbe151f2bb72526) on the terminal! Good bye to the [Unicode problem](https://github.com/NodeOS/NodeOS/issues/106)! :-P)


> I even thought about using X until the DRI platform is stable enough, but I know you are against X.

I'm not "against X" but in the fact that it add more layers that complicate the system and use more RAM + CPU resources that degrade system performance, but I have to admit it does a good job. Wayland has the same philosofy about using less layers to improve performance, the same way that DirectFB did (that's also more baremetal than Wayland), only problem is that didn't get mainstream support as Wayland has get. That's why I say about using raw DRI + EGL, because we don't need any of the extra features (windowing, transparency, network support...) but only be able to draw full-screen, and later left CEF to do the rest of the job :-)

P.D.: I'm also against fbdev because DRI + EGL can offer a better performance in a somewhat standar way, and to do the same with fbdev it's necesary to use DirectFB that internally it's using DRI so here DirectFB becames "the extra layer", but since fbdev was offered by free since it was needed to have high resolution text modes and seems it's not using too much resources and it's on the standard vanilla x86 kernel, that's why I proposed to use it as a temporal fix... ;-)

**Comment by netlovers** [comment]
I didn't change anything in my old build script:
https://gist.github.com/netlovers/0bfc342f5320f35ab601
If you check the comments the Mesa build has some instructions how to build it without any X dependency.
I know that X adds more layers, but until there isn't a browser engine that works out of the box with bare metal, I guess it would make many things easier. And graphics is an optional part of NodeOS anyway.

**Comment by piranna** [comment]
> I didn't change anything in my old build script:
> https://gist.github.com/netlovers/0bfc342f5320f35ab601

Oh great, I'll try to take a look on it.


> I know that X adds more layers, but until there isn't a browser engine that works out of the box with bare metal, I guess it would make many things easier.

There are some guys that have been able to run a QtWebKit based browser on the Raspberry Pi directly on top of DRI and framebuffer, I think we could take some ideas...

http://www.raspberrypi.org/forums/viewtopic.php?f=38&t=43087&sid=06f1b7d497ae48810a0e8ff02c6acfff
https://github.com/Topguy/mlbrowser

Anyway, although the long term purpose is to have a web rendering engine running on top of NodeOS to bring advanced graphic functionalities, what I'm saying here is to take the lower layers and start playing with them until we get the full browser working :-)


> And graphics is an optional part of NodeOS anyway.

Absolutely, and as I said in the other post, although I would love to see a desktop build in HTML and some graphic apps running on top of NodeOS, each day I think a little more if we shouldn't focus to create a pure head-less system that's only able to understand REST requests and WebSocket connections and left user interfaces (both graphic and text mode) out of scope for client apps running on external, remote devices... :-/

**Comment by piranna** [comment]
By the way, seems EGL can be statically compiled based on Gallium for a lot of platforms, that would be awesome since we would include it as a npm package... :-D

https://chromium.googlesource.com/chromium/deps/mesa/+/f2ba7591b1407a7ee9209f842c50696914dc2ded/src/gallium/targets/egl-static/Makefile

**Comment by piranna** [comment]
ChromeOS has switched to its own graphic stack Freon build directly on top of KMS/DRM to improve performance. This is also a level lower to EGL (although Ozone-* will still be available), so this are good news to us... :-)

http://www.phoronix.com/scan.php?page=news_item&px=Chrome-OS-Freon-Graphics

**Comment by piranna** [comment]
I've reviewed [topcube](https://github.com/creationix/topcube) project and seems it's mostly what we are trying to do for the GUI system, a npm package that host only a CEF-based browser and with a pure client-server architecture. It's a bit outdated and seems don't support terminal by default, but could be a great help. The needed steps would be to update its CEF version (seems it's based on Chrome 17...), compile for terminal and try to make a statically linked module.

On the other hand, Ozone-DRI backend is intented to be used directly on the framebuffer using software rendering, so it could be easier that going for pure hardware accelerated EGL... :-)

**Comment by netlovers** [comment]
I'm not even sure if Ozone-DRI is stable enough. I tried to build it multiple times the past few months, and even tried to build the Ozone-Wayland version, with no success. But this might be my fault, maybe I'm missing something important somewhere.

So if we really want a graphical interface, then we'll need someone who can build a usable Chrome version with Ozone backend, or we could use basic Chrome build on X until things get stable enough.

By the way @piranna what were your plans to make `/dev` directory (or whatever it is called) accessible?

**Comment by piranna** [comment]
> I'm not even sure if Ozone-DRI is stable enough. I tried to build it multiple times the past few months, and even tried to build the Ozone-Wayland version, with no success. But this might be my fault, maybe I'm missing something important somewhere.

Have you followed the instructions from https://www.chromium.org/developers/design-documents/ozone?  Seems should work out-of-the-box... At https://groups.google.com/a/chromium.org/forum/#!topic/ozone-dev/Q_5dkTj8Kew there are some more tips for compile it on Ubuntu, could be useful.

The Ozone Devs [seems very open to experiments](https://groups.google.com/a/chromium.org/d/msg/ozone-dev/eqV89buZjeA/oVYXjwwdX0UJ)... :-)

>> Unfortunately, X11 does not correspond to our use. 

> Are you able to share your use? We are always interested in learning who's using (or trying to use) ozone.

In the same response, they talk about a "egltest platform" for testing purposses:

> In the meantime, you could look at the "egltest" platform for a quick example that supports acceleration. Note that it is VERY limited in  functionality (no cursor, 1 surface only, fixed size).

And seems what we are trying to do is one of the intended use cases :-)

>> We are doing tests to see if we could use chromium with a specific ozone layer for STBs platforms, and at the same time having a more or less similar behaviour using ozone on a PC for debug.

> This is one of the intended uses for ozone so you ought to find this quite workable. Check out the embedded=1 target for building "embedded" content shell -- a single display-filling web contents.

:-D


> So if we really want a graphical interface, then we'll need someone who can build a usable Chrome version with Ozone backend, or we could use basic Chrome build on X until things get stable enough.

Another simpler aproach to develop it, since we want a pure client-server architecture, it's just to use a Chrome browser on a desktop PC and connect to the NodeOS server using it's IP instead of localhost as address... :-D This is a valid use case, and one of the reasons why I wanted to isolate the GUI from any Node.js specific library :-D


> By the way @piranna what were your plans to make /dev directory (or whatever it is called) accessible?

[Make a FUSE filesystem that only show the devices that are not used by anybody and the current user has permissions to access to it](https://github.com/NodeOS/NodeOS/issues/95). For example, this way several users could be able to access each one to a different /dev/fbdevX at the same time since once one of them is open the others could not be able to access to it (in fact, it will not be shown at all), or you could plug several mouses and keyboards to the same computer and each user could read only from their own and not others. Building this way a multi-seat system would be fairly trivial... :-) I've also opened an issue on [fuse-bindings](https://github.com/mafintosh/fuse-bindings) (that seems the most updated and maintained of all FUSE libraries for Node.js) to request to add the fuse_get_context() function to be able to know [who is requesting access to each file](https://github.com/mafintosh/fuse-bindings/issues/3) :-)

**Comment by derhuerst** [comment]
> Another simpler aproach to develop it, since we want a pure client-server architecture, it's just to use a Chrome browser on a desktop PC and connect to the NodeOS server using it's IP instead of localhost as address... :-D This is a valid use case, and one of the reasons why I wanted to isolate the GUI from any Node.js specific library :-D

:+1: 

Why not let NodeOS just provide an HTTP/Websockets/whatever API and host the GUI locally on the PC or somewhere else? This would make NodeOS truly platform-agnostic because native clients and even bots and CLI scripts could talk with a NodeOS instance.

**Comment by piranna** [comment]
> Why not let NodeOS just provide an HTTP/Websockets/whatever API and host
the GUI locally on the PC or somewhere else? This would make NodeOS truly
platform-agnostic because native clients and even bots and CLI scripts
could talk with a NodeOS instance.

That's the idea :-)

We are not going to enforce devs to use something beyond http/websockets to
comunicate the gui with the server but simply because they are the only
things a web browser can understand, though. At protocol layer, each one
can use whatever it wants as far both client and server understand it, only
that maybe recomend json, that's the de-facto standard.

At https://github.com/NodeOS/NodeOS/wiki/GUI is explained the diferent kind
of gui apps that NodeOS can host, but there's still the problem about how
to launch the server side of the app at the same time the client side
launch... :-/ Ideas?

**Comment by piranna** [comment]
Since applications will be on a per-user basis, I think first we need to develop the [reverse proxy](https://github.com/NodeOS/NodeOS/issues/35) so people can access to their own apps, and use http auth using logon credentials for both https and secure websockets so people can access to their own apps (it's a no-sense have credentials and later transfer data in raw. In fact, we could only allow ciphered connections). Later each user could manage what apps are running and connect to them with their own port & protocol.

**Comment by derhuerst** [comment]
>> Why not let NodeOS just provide an HTTP/Websockets/whatever API and host
the GUI locally on the PC or somewhere else? This would make NodeOS truly
platform-agnostic because native clients and even bots and CLI scripts
could talk with a NodeOS instance.
>
> That's the idea :-)

Okay, but the following sounds almost like the opposite.

> At https://github.com/NodeOS/NodeOS/wiki/GUI is explained the diferent kind of gui apps that NodeOS can host [...]

This sounds like NodeOS is going to be able to host GUI apps by itself. This is exactly what I proposed *not to do*. Can you clarify this?

> [...] but there's still the problem about how to launch the server side of the app at the same time the client side launch... :-/ Ideas?

I think, with my proposal, this problem is not relevant anymore.

> We are not going to enforce devs to use something beyond http/websockets to comunicate the gui with the server but simply because they are the only things a web browser can understand, though.

This is not true. There are dozens or even hundreds of WebSocket libraries out there, for all kinds of programming languages, platforms and use cases; C++, Objectice-C, JS, Go, etc. WebSockets are more performant than HTTP requests and offer a solid encryption.

> At protocol layer, each one can use whatever it wants as far both client and server understand it, only that maybe recomend json, that's the de-facto standard.

So you think different "desktop environments" for NodeOS are free to use different kind of transport protocols? This is possible too.

**Comment by piranna** [comment]
> This sounds like NodeOS is going to be able to host GUI apps by itself. This is exactly what I proposed not to do. Can you clarify this?

NodeOS will be able to host GUI apps by itself, just like any other web server. NodeOS GUI apps will be normal HTML + Javascript + CSS web pages that will connect to REST services and websocket endpoints on a NodeOS instance that could host all the logic of the app or only the low level systems if the logic is client side. The fact that NodeOS could have a "local" GUI is only that we are willing to exec a CEF or Gecko based web browser on a NodeOS powered machine and connect to localhost, but otherwise is a vanilla web browser where GUI will not have direcct access to the Node.js APIs at all.


> I think, with my proposal, this problem is not relevant anymore.

Yes, but it's only applicable to a desktop-like app, that can start the server process from the server side. But, how it's started this desktop-like app without wasting resources? ;-)


> WebSockets are more performant than HTTP requests and offer a solid encryption.

WebSockets are not encrypted, data is sent in raw. Only Secure WebSockets are encrypted by using TLS, the same way as HTTPS. The ones that are encrypted by default are WebRTC DataChannels, and although they are really interesting, there are no usable alternatives for Node.js (and the ```wrtc``` module seems it's not stable enough for production...).


> So you think different "desktop environments" for NodeOS are free to use different kind of transport protocols? This is possible too.

What I say it's that anything that can be done in a npm package, it's not (directly) related to NodeOS :-) In that case, a desktop in HTML could be used by a standard Node.js instance, and although we can use it, we should not enforce people to use one of them and give freedom to the user to use what it wants to enhance compatibility with current and upcoming apps.


> What I had in mind is a single, standard API for NodeOS (of course modularized and extendable) that different (user-PC-hosted) "desktop environments" can use to interact with the system.

That's one of the possibilities, but I would not enforce it the first one, although I totally like the idea of having "dumb servers" that only provides data and services and don't have application logic... In that way what I have in mind for a first iteration of the desktop GUI is being each "desktop window" an iframe and each "app window content" a HTML page. This way they can connect to localhost if they are running on a NodeOS instance or a standard OS running Node.js, or remotely from a web browser to another machine running the NodeOS instance with the app.

**Comment by derhuerst** [comment]
> NodeOS will be able to host GUI apps by itself, just like any other web server.

Why? This is not necessary at all. Furthermore, you could invest more time into the great NodeOS core instead.

You propose that NodeOS should ship with a GUI web server by default. This has a few disadvantages:

1. People who need NodeOS for non-GUI use cases: They need to live with a more complex and heavy NodeOS with features they don't need. In the worst case, there's even a web server process running by default, wasting resources like memory and CPU.
2. People who want to host the GUI by themselves. This can be out of many reasons. Latency, security or a better dev enrionment are only a few.
3. People who want to use different GUIs. Hosting two GUIs using the default NodeOS default GUI server is unnecessarily complex. Just hosting two GUIs locally on their computer is just more comfortable.

> NodeOS GUI apps will be normal HTML + Javascript + CSS web pages that will connect to REST services and websocket endpoints on a NodeOS instance that could host all the logic of the app or only the low level systems if the logic is client side.

I agree.

> The fact that NodeOS could have a "local" GUI is only that we are willing to exec a CEF or Gecko based web browser on a NodeOS powered machine and connect to localhost, but otherwise is a vanilla web browser where GUI will not have direcct access to the Node.js APIs at all.

Sorry, I'm still naively asking because I don't understand: Why? Why worry about a browser-like environment with all its implementation details if users could just use their local browsers on their PCs.

**Comment by derhuerst** [comment]
> WebSockets are not encrypted, data is sent in raw. Only Secure WebSockets are encrypted by using TLS, the same way as HTTPS.

Of course, I talked about encrypted WebSocket (WSS) connections (; But right now, the choice WebSockets vs WebRTC seems unimportant the status of this project.

**Comment by formula1** [comment]
@derhuerst When thinking selfishly, I disagree with your opinion. When thinking realistically I see warrant in what you are saying. That being said, @piranna [wrote about this sort of thing](https://github.com/NodeOS/NodeOS/issues/106)

There he basically goes over the gist of your worries. Why have the front end aspect to NodeOS packaged with and running on the server when the terminal will be the only thing being used? It doesn't make sense. He proposed the Workstation vs Mainframe concept (I did read though I didn't write anything in response, I don't want to be too big of a fanboy writing on every post of his). Additionally, because Node is mainly used for serverside programming it makes little sense to shoot for the Desktop when the server requires optimizations, user management and other extremely important features in the long run.

That being said, I also want to point out something [pretty important](https://github.com/NodeOS/NodeOS/commits/master). Since the start of the year there have been 3 committers, only 1 of which has been consistant (you can guess who). This means this project in some ways lives and dies by @piranna's motivation, which he is by no means obligated to go in a specific direction as the [pull requests](https://github.com/NodeOS/NodeOS/pulls?q=is%3Apr+is%3Aclosed) increase, the ability to satisfy requests increases.

Based off your argument, I will make the assertion that what you are suggesting has the following Pros
* Focusing on a server purposed operating system would require *arguably* less time
* Can focus on Benchmarks in comparison to other operating systems running node rather than a red herring
* more likely to be used in things like Rasberry PI and Adruino - likely where early adoption will occur
* A good server OS is more likely to become depended upon by Big Companies rather than GUI OS

However, The Idea that GUI isn't important is unrealistic
* The GUI produces Hype - I wrote some things for silk because I got really excited
* the GUI doesn't require complex knowledge of Operating System specifics and can be worked on by almost anyone.
* As the number of people using an open source technology increases, the likelyhood of people making pull requests increases (number of issues will increase as well)
* As Barrier to Entry decreases, the number of people able to try something out increases

I'm not here to tell @piranna what direction he should take NodeOS, I would much rather make pull requests to enable it. However, I haven't spent enough time working on the project due to work, OS being a scary feat to attempt as well as lower fruit getting me excited as well. I do think your opinion has value though I hope you turn your passion into a pull request rather than a reprimand.

**Comment by derhuerst** [comment]
> This means this project in some ways lives and dies by @piranna's motivation, which he is by no means obligated to go in a specific direction or even continue as he's doing it because he wants to.

Thanks for the reminder, you're absolutely right. That doesn't change my mind regarding the GUI or my personal idea of the ideal NodeOS, but it certainly is something I should respect.

**Comment by formula1** [comment]
@derhuerst That being said, the wouldn't be the Open source community if there weren't arguments or complaints :)

**Comment by derhuerst** [comment]
@formula1 What a concise summary of the pros and cons! You also brought new arguments into the discussion that I didn't have in mind.

> I'm not here to tell @piranna what direction he should take NodeOS, I would much rather make pull requests to enable it. However, I haven't spent enough time working on the project due to work, OS being a scary feat to attempt as well as lower hanging fruit getting me excited as well. I do think your opinion has value though I hope you turn your passion into a pull request rather than a reprimand.

When I stumbled upon NodeOS several months ago, I thought: "Finally, someone does it in a way that feels right!". I immeaditaly thought that NodeOS would be the perfect slim Node.js server OS. I thought of slim Docker containers, lightweight VMs running on Digital Ocean and home automation using the Raspberry Pi. **Maybe I'm a little biased fighting about this GUI stuff because I have a very server-centric view on NodeOS.**

Maybe this is is selfish, but what I'm trying to do here is to 1. make clear what my idealistic idea of NodeOS is and 2. convince @piranna that this is the way to go. :D Of course, this is a bit selfish and naive, but that is how it usually works: A discussion where people try to convince each other of their ideas.

@formula1 I'm glad to hear that you agree theoretically.

**Comment by netlovers** [comment]
The best thing about open source is that you can just press the fork button, and shape a project to be perfect for your own needs. For example I'm doing just that, because I can see that the NodeOS project is mainly going to be a server OS, and I want a desktop OS for myself, where almost everything can be controlled with web technologies. It is nice that NodeOS will have basically a remote desktop functionality, but I need more than that, so I'm working on my stuff separately.
At least this is my favorite thing about open source, and the fact that people help each other out, with new ideas, or even contribution.

**Comment by formula1** [comment]
> Maybe this is is selfish, but what I'm trying to do here is to 1. make clear what my idealistic idea of NodeOS is and 2. convince @piranna that this is the way to go. :D Of course, this is a bit selfish and naive, but that is how it usually works: A discussion where people try to convince each other of their ideas.

I like your style! lol

I was considering suggesting something like [bountyhunt](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F51989&utm_medium=issues&utm_source=github) before. I didn't end up bringing up because I really liked how this has been home grown (@netlovers for example has made big strides in embedded chromium, @zodern has been building a gui manager which I helped with, others have come in and tested/pull requests). I know nw.js has been using it to some success (though its a bit different as they are Desktop Applications which means immediate results). Perhaps this is something that can be implemented if you don't have time or energy yourself. However, to me most importantly is keeping as much stress of @piranna's and everyone else involved back.

**Comment by piranna** [comment]
> @derhuerst 
> you could invest more time into the great NodeOS core instead.

And in fact this has been always the main focus :-D If you read other issues, I have been always in favor of making NodeOS a server-only system just to improve performance and flexibility by no having a GUI at all.


> @derhuerst 
> People who need NodeOS for non-GUI use cases: They need to live with a more complex and heavy NodeOS with features they don't need. In the worst case, there's even a web server process running by default, wasting resources like memory and CPU.

That's why I proposed the concept of [flavors](https://github.com/NodeOS/NodeOS/issues/106).


> @derhuerst 
> People who want to host the GUI by themselves. This can be out of many reasons. Latency, security or a better dev enrionment are only a few.
> People who want to use different GUIs. Hosting two GUIs using the default NodeOS default GUI server is unnecessarily complex. Just hosting two GUIs locally on their computer is just more comfortable.

By using standard web technologies this is absolutelly possible. Anyway, what I propose is that each user host their own "GUI web server" inside their user home the same way NodeOS architecture promotes for everything. Remember: your home is your root...


> @derhuerst 
> Sorry, I'm still naively asking because I don't understand: Why? Why worry about a browser-like environment with all its implementation details if users could just use their local browsers on their PCs.

Think about "local GUI" as a bundled NodeOS + web browser package and connect to localhost to run your applications. In fact, you could use that bundle as a thin client to connect to the apps running on other (server-like) machine...


> @derhuerst 
> Of course, I talked about encrypted WebSocket (WSS) connections (;

Ah, cool then :-)


> @derhuerst 
> I have a very server-centric view on NodeOS

Me too. As @formula1 pointed out GUI would be some kind of a hype and sometimes I think could be useless, but it could be funny too. NodeOS is a server OS, period. The point that people want a GUI only means that we need to think about how to do it well if we want to success, and this means don't loose this point. That's it, GUI will be client apps running on a web browser, only that we'll "fake" the GUI making that web browser running on the same machine where the server is. This is similar to how X11 works internally, everybody use local desktop linux OSes but internally it's a client-server architecture.


> @netlovers 
> It is nice that NodeOS will have basically a remote desktop functionality, but I need more than that, so I'm working on my stuff separately.

Could you explain your use case and why you are doing it that way? Just curious, but maybe your changes could be integrated as well :-)

> @formula1 
> However, to me most importantly is keeping as much stress of @piranna's and everyone else involved back.

Oh, thank you :'-) By the way, you are not having news from me the last weeks because I have been trying to write the bachelor thesis (that I didn't finished yet... :-( ) and will need to study for my examns, I'll try to come back on July :-)

**Comment by piranna** [comment]
By the way, I sometimes has problems to explain concepts that are simple and clear on my head but people don't end to understand, and this also happens (in fact, specially) at my workplace. Sorry for the inconvenience.

**Comment by formula1** [comment]
@piranna Flowcharts are a universal language :P (And imo prettier than english)

**Comment by piranna** [comment]
Are you suggesting I should do a flowchart of how NodeOS GUI should work (and by the way explain thinks at work with them)? Maybe you are right and I should start doing them to improve documentation... Any suggestion better than Dia? I love it but sometimes gets tedious...

**Comment by piranna** [comment]
https://github.com/knsv/mermaid

Ok, that seems cool :-) Unluckily markdown don't accept plugins like MediaWiki... :-(

**Comment by formula1** [comment]
I generally do flow charts to explain things at work, also for myself. All too often I'll get stuck trying to think in english (for you in spanish most likely). However, the more I think in english, the more I find myself forgetting the basics of what I'm doing and get too deep.

Github disallows Scripts, however, [it does render svgs](https://github.com/specialorange/FDXCM/blob/master/doc/models_brief.svg) which D3 produces. Perhaps, in a seperate repository, the flowcharts can be stored. I doubt you'll be writing flowcharts too often except for the important things, though it would be nice to just do it on the fly : /.

**Comment by piranna** [comment]
> though it would be nice to just do it on the fly

Yep,specially on the issues :-/

**Comment by netlovers** [comment]
This looks interesting: https://github.com/citronneur/node-rdpjs
And if you check the example implementation: https://github.com/citronneur/mstsc.js it uses canvas to render the screen.

**Comment by piranna** [comment]
Now that node-canvas is working on NodeOS, this can be a cool proof of concept to use it as a thin client :-)

**Comment by piranna** [comment]
Support for FbDev is finished in node-canvas both on real hardware and QEmu, now wait it gets merged... :-P By the moment you can use it directly from the NodeOS fork repo :-)

**Comment by formula1** [comment]
That's really exciting, it's really awesome everything that is happening here

**Comment by piranna** [comment]
Don't stop the music ;-)

**Comment by zreo15** [comment]
how about a browsers rendering engine written purely in js... it can run pretty easily on top of node.js js engine u can easily link the OS apis to be accesed by local hosted webpages as GUI for everything 
https://github.com/trevorlinton/webkit.js try it should work in theory just like nw.js or atom but dosent exposes node apiz like they do

**Comment by piranna** [comment]
It could be possible, but it would be slow, too. Also, there's the problem about how to output the content. Does webkit.js support to output to a canvas object?

**Comment by netlovers** [comment]
@piranna actually it only supports canvas. You can check out the demo here:
http://trevorlinton.github.io/

**Comment by piranna** [comment]
Didn't know it... Then yes, it could be an alternative, so if anybody want to experiment with it I can help... :-) The idea would be to use webkit.js against NodeOS fork of node-canvas using FbDev as backend, so should be easy :-)

**Comment by piranna** [comment]
Ok, and create a pre-generated npm package, seems it's dificult to generate...

**Comment by netlovers** [comment]
I guess one should be able to use the files in the bin directory for testing purposes. After it is confirmed that it works, maybe we can think about a better way to package it.

**Comment by piranna** [comment]
I agree. Seems it works with limitations, but on the mail list they talk about problems on Node.js... But the demo seems easy to use, so seems doable to craft an script to render an url given as argument that combine webkit.js and node-canvas as dependencies.

**Comment by zreo15** [comment]
well since the code base is js so it will be faster then conventional rendering and the problem on running on top of node.js I encountered it maybe its something in my system someone else should try to run should work well

**Comment by piranna** [comment]
No, it will not be faster, in fact it will be a lot slower not only by using Javascript instead of compiled binary but also because it will not have hardware aceleration at all, but anyway it could be a good and easy task to check out.

**Comment by zreo15** [comment]
well can you test it on windows and linux first to see if its workin i cant seem to get it to work on mine and please post ur instructions too on how you did it

**Comment by piranna** [comment]
Just for testing, I think it should be mostly enough to pick the generated file directly from the browser demo and init the constructor with the canvas object generated by node-canvas...

**Comment by zreo15** [comment]
umm all this is above my head can you run your test and then post the instructions for newbie

**Comment by piranna** [comment]
I'm not in my laptop and don't have too much time, sorry. Anyway, this quick test should easy also for a newbie, mostly copy&paste

**Comment by zreo15** [comment]
ohk then please give me instructions what do I have to do will try

**Comment by piranna** [comment]
I've review the code and seems it's too much focused on desktop, so it will need to be ported to Node.js. I recomend you to start downloading the project and generating the code for the browser, and later changuing https://github.com/trevorlinton/webkit.js/blob/master/src/webkit.api.js to make it work in Node.js.

**Comment by zreo15** [comment]
ya I had problems in my system too though will see how it goes in other sysytem and try to build it again and find a way to get it working with node.js 
but any idea how to use it to open a browser window using webkit.js running on node.js

**Comment by piranna** [comment]
Do you want to open a browser window to exec there webkit.js? Forget about it, it will be a waste of time.

**Comment by zreo15** [comment]
sorry i meant like a window to display content that will be rendered 
if no window is displayed then running this on node will be same as running phantom js like headless webkit

**Comment by netlovers** [comment]
Render the content to a png file, or something. If that works, it should be easy to change it to render to the linux framebuffer.

**Comment by zreo15** [comment]
well theres no api for screen shot and I have no idea how to remove the restrictions from the demo.html so cant test if this works on whole website but I checked svg and its working but canvas element not since m guessing the demo already displays the content on a canvas element

**Comment by formula1** [comment]
@zreo15 I like your passion, perhaps you need a little direction.

We cannot do screenshots, not do we have a webbrowser to actually load the webkit.js to. This makes things seemly impossible however, it is not so.

* This is how [webkit.js is initialized](https://github.com/trevorlinton/webkit.js/blob/master/bin/demo.html#L108) - created as `new WebKit(width, height, canvas, scale, hidpi)`. hidpi can probably be 0, scale 1. But most importantly is canvas, width and height
* Normally we don't have access to a `canvas` element context. However, a library like [`node-canvas`](https://github.com/Automattic/node-canvas) provides that sort of thing for us
* From here we make our best attempts to make webkit.js available in node. 
  * adding to the bottom of the compiled file `module.exports = WebKit`
  * creating a fake `var window = {};` at the top
  * Changing line [149](https://github.com/trevorlinton/webkit.js/blob/master/bin/webkit.js#L149) to include `|| setImmediate`
  * utilizing [a module](https://www.npmjs.com/package/workerjs) a creating a wrapper around child_process api to shim webworkers. Line [20](https://github.com/trevorlinton/webkit.js/blob/master/bin/webkit.js#L20) is the important one here
  * Try to remove any reference to document like here on line [224](https://github.com/trevorlinton/webkit.js/blob/master/bin/webkit.js#L224)
* Test by implementing what they have in the demo.html only within a `render.js`

The important code seems to be around line [107](https://github.com/trevorlinton/webkit.js/blob/master/bin/demo.html#L107) in demo.html

```
        window.webkit = new WebKit({
          width:document.getElementById('width').value,
          height:document.getElementById('height').value,
          canvas:document.getElementById('canvas'),
          scale:document.getElementById('scale').value,
          hidpi:(document.getElementById('hidpi').checked ? 1 : 0)
        });
        window.webkit.addEventListener('call',function(e){
          running[e] = typeof(performance) != 'undefined' ? performance.now() : Date.now();
        });
        window.webkit.addEventListener('done',function(e){
          var t0 = typeof(performance) != 'undefined' ? performance.now() : Date.now();
          document.getElementById('info').innerHTML += e + ' took ' + 
              (Math.floor(((t0 - running[e])/1000)*1000)/1000) + ' seconds.<br>';
          running[e] = null;
          if(e == 'setHTML' || e == 'setHtml') {
            document.getElementById('render').value = 'Render';
            document.getElementById('render').disabled = false;
          }
        });
        window.webkit.addEventListener('ready',function() {
          document.getElementById('info').innerHTML += 'Running...<br>';
          window.webkit.html = document.getElementById('htmltorender').value;
        });
```

If that is too much, its ok. I plan on going through this when I get the chance

**Comment by piranna** [comment]
> Changing line 149 to include || setImmediate

Use setImmediate() to mimic requestAnimationFrame() would be criminal, it will suck all the CPU. Way better to use setTimeout(func, 1000/60) like does the polyfills (or maybe better, remote that like at all and use directly a polyfill :-) ).


> utilizing a module or creating a wrapper around child_process api to shim webworkers. Line 20 is the important one here

I think for first rendering tests we can use it with aceleration disabled.


> Try to remove any reference to document like here on line 224

They can be changed by a method where to pass the event to by processed by webkit.js, and move the fetch the real events to the demo or the Node.js program instead.

**Comment by zreo15** [comment]
yes for past few days i was trying to understand these things but they are all very high level for me anyway thanks for clarifying things will try to learn more and experiment 
Yes i want to try working on this I will tell when i hav some success in anything

**Comment by netlovers** [comment]
I tried to make it work, however in a VM with 512 MB RAM it always killed itself. It kind of works with 1024 MB. But it looks like although it uses the canvas element I passed to it, for some reason it calls WebGL methods on it. And it looks like `node-canvas` does not support WebGL calls.
For example one missing method is `texSubImage2D`, and there is a bunch of these. If somebody wants my hacky code then I could put a patch to a gist. But as far as I can tell this won't work. At least not with `node-canvas`.

**Comment by zreo15** [comment]
how about you disable webgl and build webkit.bin.js and try the tests

**Comment by netlovers** [comment]
I don't think that's possible. It uses WebGL calls internally, and I haven't seen fallbacks for them.

**Comment by zreo15** [comment]
well in features.gyp the webgl seems to be turned on try disabling that and building and WebGL calls should be internal and only canvas readable data be sent to node-canvas I guess then how is webGL getting calls from node-canvas

**Comment by piranna** [comment]
> however in a VM with 512 MB RAM it always killed itself. It kind of works with 1024 MB

That sound really bad, too much memory... :-/


> node-canvas does not support WebGL calls

node-canvas only support 2D context since it's mostly focused for off-line rendering. It should be feasable to add support for 3D graphics now that it start to have support for real hardware (in part thanks to NodeOS :-) ), but it will need to be done from scratch, so not soon :-( Also, first would be needed to have acelerated graphics... :-/


> It uses WebGL calls internally, and I haven't seen fallbacks for them.

I believed had seen an option for this, but don't find it now... :-(

**Comment by piranna** [comment]
> well in features.gyp the webgl seems to be turned on try disabling that and building and WebGL calls should be internal and only canvas readable data be sent to node-canvas I guess then how is webGL getting calls from node-canvas

https://github.com/trevorlinton/webkit.js/blob/15306b5c540ccafc3be008225eafebd1932d356d/tools/features.gypi#L119

Holy sh*t! 8-|

**Comment by netlovers** [comment]
I don't think turning it off helps. If I'm correct it is used to render WebGL inside the main canvas, and I don't think it would change the current behavior. But you can try if you want.

**Comment by piranna** [comment]
Do you think OpenGL is being used for display rendering instead of only for WebGL? Anyway, with so much memory needs, seems not to practical for real use anyway...

**Comment by netlovers** [comment]
Yes, I think so. I knew about WebGL, but never wrote any code for it. So now I checked, and it renders to canvas. So in the `webkit.js` project's description this is still true:
```
supports rendering to WebGL/Canvas contexts
```

**Comment by piranna** [comment]
:+1:, but I think that statement is more related to support WebGL/Canvas instead of render to it...

**Comment by zreo15** [comment]
i think by disabling some more features memory consuption can be reduced and can you post the build and your test

**Comment by netlovers** [comment]
Yeah, use these commands:
```bash
mkdir webkit-test
cd webkit-test
wget https://raw.githubusercontent.com/trevorlinton/webkit.js/master/bin/webkit.bin.js
wget https://gist.githubusercontent.com/netlovers/7fc6c304d92fe09277d2/raw/c57c52dd1b42edec6c59d14f2522c1a6ed8809ff/webkit.bin.js.patch
patch < webkit.bin.js.patch
npm install canvas
```

And then run:
```
node index.js
```

This will create a `webkit-test` directory with all the necessary stuff to run `webkit.js` in node. I left some `console.log`s in it for easier debugging.

**Comment by piranna** [comment]
[A pure Javascript implementation of WebGL painting on a Canvas 2D element](http://cimaron.net/projects/cwebgl/). Crazy, and webkit.js would be boring slow, but what the f*ck :-P

**Comment by piranna** [comment]
And maybe also [html2canvas](https://github.com/niklasvh/html2canvas) would be used to "paint" web page on node-canvas, but it would need more work that could be used towards making content_shell to work...

**Comment by netlovers** [comment]
`html2canvas` is not suitable for nodejs use.

**Comment by zreo15** [comment]
hey after screwing my ubunutu for password for whole day i got your example working but itz not displaying anything just printing some console log

**Comment by netlovers** [comment]
Yes, because that is the only thing it does. If you wait a few seconds you can see that it created a longer `data:...` line, you can view that in your browser, but it is only a white or transparent 500x500 image. That `data:...` stuff should be the rendered webpage, if everything works correctly.

**Comment by zreo15** [comment]
ohk got it will start experimenting with your example

**Comment by formula1** [comment]
Something worth noting but difficult to implement is [headless-gl](https://github.com/stackgl/headless-gl)

**Comment by lite20** [comment]
I'm paying extra attention to try and make sure this GUI works with the graphic implementation for NodeOS. Is this a full Desktop Environment? (Login/Desktop manager, Window Manager, Navigation interface and so on and so forth). From the images I only see the Desktop (without the login) and potentially window management? 

Looks good, keep it up :smile:

**Comment by zodern** [comment]
In NodeOS, the logon script is run as root, and then there is no more root user. Silk would have to run instead or along with the logon script to handle logging in. More on my idea for this below.

Those screenshots are old and there has been a lot of progress since that first post. There is a simple window manager. I never head of a desktop manager before (I have a lot to learn), but if I understand what it is I have it partially created but there is no file picker yet and many other things are missing. There is now a few built in apps: Terminal, File Explorer, App Manager, Task Manager, Text Editor and a server only app for app defaults.

 Eventually there will be 3 different ways silk can be used: 

1) As a full desktop environment. It will start when the computer starts and run until it is shut down. This might require a fork of NodeOS, but that could be automated to keep it up to date with NodeOS. There would probably be a small app running as root for logging in and then once the user logs in it will launch Silk.
2) A graphical way to access a server. Currently you can install it on the server, run with `silk -r` and go to the remote url. It needs to be more secure and not have to be started each time using ssh.
3) Using a computer (probably NodeOS though it will work on any OS) and decide you want to use a GUI for something. `silk -o` starts silk and then opens nw.js which loads silk in an iframe with node disabled.

What do you mean by navigation interface?

**Comment by zodern** [comment]
I forgot the main use: 
4) Have a way that a user can choose to have Silk start when they log into NodeOS. That way each user can decide which GUI to use or none at all. Also, NodeOS will take care of logging in.

 # 4 is the most important, # 3 is needed for easy development and to allow people to easily try Silk, and # 2 I think is a really cool possibility. # 1 is the end goal.

**Comment by lite20** [comment]
> I never head of a desktop manager before (I have a lot to learn), but if I understand what it is I have it partially created but there is no file picker yet and many other things are missing. 

When you start your computer, the display manager is what shows the login screen. Once signed in, the display launcher launches your Desktop Manager, and Window manager. The window manager handles windows, and the Desktop manager handles the desktop (files on the desktop if there are any in that design, along with the background wallpaper and such). By navigation I been things like dock-lets. For example, the dock on the side in Ubuntu. There are an infinite amount of possible ways to navigate through a computer, and each desktop environment tries to unleash a different method. 

All these put together are what we call a desktop environment. I hope that sheds a bit of light! :smile:

**Comment by zodern** [comment]
Thanks. That is a lot more clear than what I found on Google.

Once a user logs into NodeOS, there probably is no need for the display manager except when accessing Silk remotely. If there is different flavors of NodeOS as suggested in #106, the workstation flavor could have an html login screen and then launch whatever GUI if any the user wants.

There is a window manager. There is not really a desktop manager. The desktop is just a css gradient. 

For the navigation interface it currently is similar to Windows but I want to try something different.

**Comment by lite20** [comment]
> For the navigation interface it currently is similar to Windows but I want to try something different.

I'm very curious to see what you create. I too of course will be working on a desktop environment, and by the looks of things in #178 @piranna has an idea too of what he thinks it should look like so we may have multiple environments for users to choose from which is perfect.

Mine is a mobile first implementation for reasons also outlined in #178 and @piranna and you seem to have desktop implentations. If we wanted to ship them all officially, then in the "installer" (more of a configurer because NodeOS contains all it's files being so small) we could let the user select. And of course, being designed the way NodeOS is (beautifully), users can easily install any desktop environment, or components of desktop environments via npm with ease.

Everyone wins :stuck_out_tongue: 

Cheers! Awaiting feedback :)

**Comment by piranna** [comment]
> In NodeOS, the logon script is run as root, and then there is no more root user. Silk would have to run instead or along with the logon script to handle logging in.

In traditional systems the login is done by a Desktop Manager (GDM for Gnome, KDM for KDE, LightDM on Ubuntu...), that's mostly what's `logon` is doing now on the terminal. I think for the HTML GUI it would be just enough to use HTTP Auth, and would be reused for `wssh` too. This way the "desktop environments" (Gnome, KDE, Unity...) are installed, managed and booted in a per user basis, not globally. I think they could be client-side HTML+Javascript webapps and served by a static web server inside NodeOS or other place, and also they could be installed locally on the user browser and connect via websockets to the NodeOS apps or maybe something as simple as WebDAV or HTTP requests to use NodeOS only to store the files. There's no need to install anything globally, please try to think this way, no global or shared elements ;-)


> There is now a few built in apps: Terminal, File Explorer, App Manager, Task Manager, Text Editor and a server only app for app defaults.

Where do you have the code of these applications to take a look on them?


> 1) As a full desktop environment. It will start when the computer starts and run until it is shut down. This might require a fork of NodeOS, but that could be automated to keep it up to date with NodeOS. There would probably be a small app running as root which would show an html page for logging in and then once the user logs in it will launch Silk.

I agree, that login would be just static content and redirect to the user desktop as an alternative for the HTTP Auth (while still using it), so it would still be secure. We need to develop the reverse proxy to the users folders, so this login page could be integrated or configured to be shown when accessing to the root page, +1.


> 2) A graphical way to access a server. Currently you can install it on the server, run with silk -r and go to the remote url. It needs to be more secure and not have to be started each time using ssh.

What's doing that server at this moment?


> 4) Have a way that a user can choose to have Silk start when they log into NodeOS. That way each user can decide which GUI to use or none at all. Also, NodeOS will take care of logging in.

Yes, on local access, when we get `content_shell` working it would be configured as the user's shell, so when he logs from the terminal it get directly the GUI, I like the idea :-)


> @piranna and you seem to have desktop implentations

Mine would be somewhat easily adapted for a mobile screen just by removing the split views and having only one type of window... ;-) But most important: if we attach ourselves to the apps being regular web pages inside iframes and use standard W3C technologies, they could be easily shared between the different desktops, so let's have this in mind :-)

**Comment by lite20** [comment]
> Mine would be somewhat easily adapted for a mobile screen just by removing the split views and having only one type of window

Oh okay, that's pretty awesome!

> But most important: if we attach ourselves to the apps being regular web pages inside iframes and use standard W3C technologies, they could be easily shared between the different desktops, so let's have this in mind :-)

Definitely. The worse thing we could do to users is lock them into a desktop environment because certain apps are dependant on it's features.

Sounds good!

**Comment by zodern** [comment]
> Where do you have the code of these applications to take a look on them?

Most are in the [Silk-GUI organization](https://github.com/Silk-GUI) and there is @formula1  [Tunnel Client](https://github.com/formula1/Silk-Tunnel-Client) for a remote desktop like feature.

> What's doing that server at this moment?

`silk --remote` currently is using [local tunnel](https://github.com/Silk-GUI/Silk/blob/master/core/remote.js#L36). The Tunnel Client app made by @formula1 uses his [Silk Server](https://github.com/formula1/Silk-Server), which I will be switching to.

> The worse thing we could do to users is lock them into a desktop environment because certain apps are dependent on it's features.

A lot has been written about making the front end be regular web pages to work cross browser. I am wondering about the node.js server most apps will have. Currently Silk does some things for apps on the server:
- Routing to make the public folder and bower components accessible on localhost
- Allow calling methods on the server from the client, and then listen to updates sent by the server
- Manages app's bower and npm dependencies

I am hoping to make each of those things optional and move some of  them to modules published on npm which apps can depend on.

There is also an api for interacting with Silk. 

Besides those points, it is plain web apps. Eventually, node.js apps like Node Music Player will work by adding an app.json and then optionally adding code for opening files or other Silk specific features. The terminal, which uses a fork of [tty.js](https://github.com/chjj/tty.js), almost works that way. The app requires the forked tty.js and then starts tty.js server. Since the tty.js server uses it's own port it does not work remotely yet.  

What I am wondering is how much should the desktop manager do for apps and what needs to be done to allow the server part of apps to work across desktop managers? It seems it would involve a standard API all of the desktop managers support.

**Comment by lite20** [comment]
> Eventually, node.js apps like Node Music Player will work by adding an app.json and then optionally adding code for opening files or other Silk specific features.

If you make a generic interface, I will implement the same. (Woohoo! You get to write a spec! :laughing: )

> What I am wondering is how much should the desktop manager do for apps and what needs to be done to allow the server part of apps to work across desktop managers? It seems it would involve a standard API all of the desktop managers support.

This is something you should probably make a new issue for so we can discuss it in depth. That's a very good point you're bringing up. I believe simply providing/assigning a surface for the apps to render to should suffice. Of course each desktop manager should expose an API for manipulating it's unique features, but by providing a basic outline of the mandatory things, that remain the same, apps will be able to operate properly under any desktop manager, and anything accessing API specific features was obviously designed for that desktop manager. 

That way nothing is dependant. It's simply optional :stuck_out_tongue:  (if that makes any sense. Sorry, its late)

**Comment by piranna** [comment]
I think we'll need to develop ASAP the [reverse proxy](https://github.com/NodeOS/NodeOS/issues/35) so we could redirect petitions to each user daemon server, so we can be able to don't need a central server nor need to invent our own protocols (I think for a GUI app NodeOS would work mostly as a storage pod, doing all operations on the client).

I think we could go here the `logon` way, detecting the reverse proxy the incoming connections and starting the server defined by the user the same way `logon` start a shell instance, so this way the user don't need to have a web server running all the time (except if they wants, so they could start it from its `init` and register on the main one :-) ). For WebSockets and keep-alive connections the server could be running without problems until all of them gets clossed, problem would be to "one shoot" requests, but maybe it could be designed to reuse one of the previous instances used by the WebSockets or the keep-alive, and also maybe add a timeout of some seconds to don't close inmediatly so it can be reused if new connections gets received.

Hum, seems to me to be an interesting and funny project now that I've found how to solve the problem of keep the servers up and running... :-D

**Comment by piranna** [comment]
By using this reverse proxy scheme apps would have the next advantages:

* each user would have it's own desktop environment (yay! :-D)
* terminal apps would be just a WebSocket connection to localhost (when running on local)
* access to files will be plain HTTP requests (GET, PUT, POST, PATCH...)

To register permanend web servers, it would be as simple as to use a patched `http` module that send the request to a UNIX socket where the reverse proxy will be listening. Fairly simple to use all :-)

**Comment by lite20** [comment]
@piranna did you post in the wrong thread or am I just that lost?

**Comment by piranna** [comment]
The reverse proxy would allow to each user to run its own web server and access to it from the external world, and this can be used to access to its data and to its desktop :-)

**Comment by piranna** [comment]
https://github.com/piranna/oneshoot

Et voilá! The per-user WebSockets and static web server. This is supposed to be installed by each user and accept the WebSocket and HTTP petitions from the GUI desktops, in the next days I'll develop the reverse proxy to start them in a secure environment and redirect that WebSocket and HTTP petitions to their correct one. Main problem will be to coordinate the transference of the free ports, but maybe I would be able to use a pipe or a domain socket, so this would have the advantage of not needing more external ports :-)

**Comment by piranna** [comment]
Ladies and gentleman: it has give me a lot of problems, but now the [nodeos-reverse-proxy](https://github.com/piranna/nodeos-reverse-proxy) is working! :-D This will process request based on the url name field (only HTTP Basic Auth by the moment, but shouldn't be too difficult to change it to Digest) and redirect to the per-user instance of `onsshoot`, starting it if there's no one currently working. It's freaking awesome to see the user directory content as a JSON array from Firefox on my laptop while NodeOS is running inside QEmu... :-D

Both are now included on the pre-generated images, so now the HTML5 GUIs can use them and allow to each user to have its own desktop. @zodern, @lite20, @netlovers, @formula1 and the others: could you be able to check it out? `oneshoot` can be run standalone so you can test it on your machines, to disable the timeout just exec it with `--timeout 0`.

Now we can build something BIG here!!! :-D

**Comment by lite20** [comment]
> Both are now included on the pre-generated images, so now the HTML5 GUIs can use them and allow to each user to have its own desktop. @zodern, @lite20, @netlovers, @formula1 and the others: could you be able to check it out? oneshoot can be run standalone so you can test it on your machines, to disable the timeout just exec it with --timeout 0.

I'm really excited for this! The future is now!

> Now we can build something BIG here!!! :-D

Yeah! What ports are open by default on NodeOS?

**Comment by piranna** [comment]
> What ports are open by default on NodeOS?

At this moment all of them, the firewall is not configured, but we should
do it in the future. QEmu is configured to access ports 50080->80 and
50433->433, but the reverse proxy is only listening to the 80.

**Comment by lite20** [comment]
> QEmu is configured to access ports 50080->80 and
50433->433, but the reverse proxy is only listening to the 80.

Ah okay. I was just having issues connecting to oneshoot running on QEMU. Does ifconfig in NodeOS return the correct IP or is that a default of sorts?

**Comment by piranna** [comment]
> Does ifconfig in NodeOS return the correct IP or is that a default of sorts?

Inside QEmu it will always get the `10.0.2.15` by DHCP. Don't know how complete is NodeOS `ifconfig`, I think it only has support to put up the network interface, but it's not being used at this moment since at this moment the IP is being get by the kernel itself on the boot process.

**Comment by zodern** [comment]
> If you make a generic interface, I will implement the same. (Woohoo! You get to write a spec! :laughing: )

Let me get it working, and then we can discuss this. Basically there needs to be a way to pass which port to use to the app (maybe as an argument when spawning) or allow the app to share the port with the desktop, and then allow calling the api across processes. Would your "voidspace" help with that?

>  by providing a basic outline of the mandatory things, that remain the same, apps will be able to operate properly under any desktop manager

Interacting with windows and opening a file is what first comes to my mind as being part of a standard api. Should we also have a standard app.json for apps? I will create the issue next week when I get back from my trip (I won't have internet for the rest of the week). 

@piranna That is exciting! I got it to work. It would be nice if it would let you try again after getting the password wrong :smile: . I switched to private browsing to get around that. Also, I am confused if oneshoot is a GUI or if it will be used by GUIs because of this [line](https://github.com/NodeOS/NodeOS/blob/9902f22f1d3a77aee40920f1d99b5b1610fb84b1/node_modules/nodeos-usersfs/logon.json#L4). 

I got silk remote to work while running in NodeOS, though only some of the apps started within the allotted 5 seconds.

**Comment by piranna** [comment]
> Basically there needs to be a way to pass which port to use to the app (maybe as an argument when spawning) or allow the app to share the port with the desktop, and then allow calling the api across processes

At this moment `oneshoot` is writting the port that's using on `stdout` so this way `nodeos-reverse-proxy` can know where to target the requests. I was thinking the same mechanism would be used for GUI apps that need a custom server backend: use the `oneshoot` WebSocket to start a new instance of the server, capture the port where it's running, and open the new window pointing to it. I was thinking maybe we could do it in a generic way, for example using a new file descriptor instead of stdout to don't clash with apps current behaviour.


> Interacting with windows and opening a file is what first comes to my mind as being part of a standard api.

Interacting with windows is more a task of the Windows Manager, on the client side. Do you propose to have here a common API to do common tasks (maximize, minimize, close, move...)? I think OpenDesktop has something similar, maybe we could take a look there. Regarding files access, the simple and obvious thing is use directly HTTP methods.


> Should we also have a standard app.json for apps?

I think it's better to re-use `package.json` adding the custom fields that we would need (icon? anything more?) instead of add a new manifest file...


> It would be nice if it would let you try again after getting the password wrong :smile: .

I'm not sure if this is how HTTP Auth works, maybe there's a way to revoke the credentials...


> Also, I am confused if oneshoot is a GUI or if it will be used by GUIs because of this line.

It's intended to be used by GUIs to do standard tasks like serve static content and provide WebSocket connections, if a GUI app needs to access to some Node.js APIs because they are not available on current browsers (for example open an UDP connection or configure the wifi or mount a pendrive) then it would use a WebSocket connection to start it's own backend server to do this non-standard tasks. I know, this would lead to have two OSes at the same time with incompatible APIs (NodeOS in Node.js, and NodeOS GUI in HTML + Javascript), but it's the best option for organization and interoperability.


> I got silk remote to work while running in NodeOS, though only some of the apps started within the allotted 5 seconds.

That's strange, the 5 seconds only take effect when there's no more open connections to that user server (both HTTP and WebSocket) and get cancelled and renewed when a new one occurs... I've just reviewed the code and didn't find any problem here... :-/

**Comment by piranna** [comment]
I have been thinking regarding the GUI apps, and I think we can make the scheme really simple. We have two kind of apps:

* pure web apps, that can run on regular browsers
* hybrid web apps, that require a server to access Node.js APIs

Apps can be "stored" inside npm packages, augmenting the `package.json` file with some fields like `icon` and others, and pure web apps could set the `main` field to its index.html. Hybrid web apps would be regular Node.js apps with its own webserver and so, being that at `server.js` (the default for `npm start`).

We launching a GUI app, to identify them it's just a matter of check for the existence of the `icon` field on `package.json` and a `index.html` or a `server.js` file. Pure webapps will just open the `index.html` file by its URL, while hybrid ones will need to start the `server.js` first (HTTP REST request? WebSocket?), that will return the URL of the just started web server.

In that sense, `oneshoot` would be considered the `server.js` of a Desktop Environment app (or something very similar) but this has a drawback when connecting from an external machine (at this moment the only way), that's required to NAT a port for each app. HTTP Digest Auth has a `realm` field that seems to be somewhat specific to the resource, so it would be set to the TCP port of the app so `nodeos-reverse-proxy` can be able to identify it and do the redirection by only exposing the ports 80 and 433, the same way currently is being done with the user name.

Thoughts? Comments? Do you think there's another way?

**Comment by piranna** [comment]
I've reviewed the Auth Digest and it's not possible to fetch the realm on the server. I thought about sending the port as a part of the path to identify the requests that don't go to the main one (the GUI server, oneshoot), but this has the problem that's not possible to know when a server can be clossed or how to restart it, so only solution seems to be to send a custom header with the path of server.js file so nodeos-reverse-proxy can be able to start it on demand. This has the advantage that alloe to make the system fully homogeneous, too :-)

**Comment by lite20** [comment]
> Basically there needs to be a way to pass which port to use to the app (maybe as an argument when spawning) or allow the app to share the port with the desktop, and then allow calling the api across processes. Would your "voidspace" help with that?

Voidspace helps with sharing variables across multiple processes. It also allows RPC calls to be made between them. If that's what your program needs, yes voidspace can help :smile: 

Voidspace is accessible externally, as long as the port voidspace is listening on, is known by the connecting client, and open for connections.

**Comment by lite20** [comment]
> solution seems to be to send a custom header with the path of server.js file so nodeos-reverse-proxy can be able to start it on demand. This has the advantage that alloe to make the system fully homogeneous, too :-)

This doesn't sound toooo hard :stuck_out_tongue_closed_eyes:

**Comment by piranna** [comment]
> This doesn't sound toooo hard :stuck_out_tongue_closed_eyes:

Yeah, but there's the problem of needing to add the header on the petitions, and this is not transparent... :-/ Honestly, the only problem here is when an app makes use of AJAX requests or asynchronous load (like by using templates or specially injecting img tags or similar, where you can't be able to add headers), because if not, all the request would go the server could be done on load and the server would die after that, so the problem is about how to restart it later :-/

I think I'll try to do something working to try to find how to make this work by using a "design by doing" approach, I think static content would be served by the main one without problems.

**Comment by lite20** [comment]
> when an app makes use of AJAX requests or asynchronous load (like by using templates or specially injecting img tags or similar, where you can't be able to add headers)

Ahh I see what you mean. What was the issue again with sending the destination as a URL parameter?

**Comment by piranna** [comment]
> What was the issue again with sending the destination as a URL parameter?

To be able to route it to the port where the app server is listening. I was thinking about using the port as a prefix in the URL, but this could change. Now I think I could be able to use the server path as the prefix the same way I was going to send it on the header, since this doesn't change. I would need to parse the URL, but maybe it's possible...

**Comment by lite20** [comment]
> To be able to route it to the port where the app server is listening. I was thinking about using the port as a prefix in the URL, but this could change. Now I think I could be able to use the server path as the prefix the same way I was going to send it on the header, since this doesn't change. I would need to parse the URL, but maybe it's possible...

Oh so that is what you're going to do? Okay yeah that sounds like an excellent idea. If you need a hand, tag me as always ^_^

**Comment by piranna** [comment]
> Oh so that is what you're going to do? Okay yeah that sounds like an excellent idea. If you need a hand, tag me as always ^_^

After thinking too much, the only solution I have found to be able to map per-application servers to the external world is to include the server name or path inside the URL, separating it with some special charactes. In this way, I think the best is to identify the first path element to see if it's in brackets, and if so, its content say the server that must be used. For example, `ws://12.34.56.78/[terminal]/foo/bar` would start the server of the `terminal` app on the `/foo/bar` directory. The brackets would host a full path if the server is not from a (user) globally installed module, like `http://12.34.56.78/[/Steam/steam-ui]/hl3` would start the web server located at `/Steam/steam-ui` (if it's not already started) and give him the path `/hl3`.

@lite20 and @zodern, what do you think about this scheme to start backend servers for the GUI apps?

**Comment by zodern** [comment]
I like that idea. 

Above I wrote

> 2) A graphical way to access a server. Currently you can install it on the server, run with silk -r and go to the remote url. It needs to be more secure and not have to be started each time using ssh.

If silk could be started with ws://12.34.56.78/[silk]/foo/bar that would be great. Go to `remote.silkgui.com`, enter the server address and then get redirected to the window manager once it is ready.

When accessing icon.png from the steam-ui server after it has been started, would I use `http://12.34.56.78/[/Steam/steam-ui]/hl3/icon.png` or `http://12.34.56.78/hl3/icon.png` or something else ?

For the actual apps, I think it would be better for Silk (or the desktop the user is using) to start them so it can make the api available to the app and have control over the process so it can stop the app when the user is done with it. Though if I use VoidSpace I could use that instead for the api.

**Comment by zodern** [comment]
One concern I have is Silk will have one way of running when in NodeOS and another when running on any other operating system.

**Comment by piranna** [comment]
> If silk could be started with ws://12.34.56.78/[silk]/foo/bar that would
be great.

That's just the idea :-) nodeos-reverse-proxy will then later request
itself the credentials using HTTP Auth.

> When accessing icon.png from the steam-ui server after it has been
started, would I use http://12.34.56.78/[/Steam/steam-ui]/hl3/icon.png or
http://12.34.56.78/hl3/icon.png or something else ?

Good quedtion, since this was the use case that made me bang the head
against the wall all the weekend. It should be used relative paths, it's
said, hl3/icon.png, this way it gets transparent to the brackets section in
the url both if using its own web server or by using the static one from
oneshoot.

> For the actual apps, I think it would be better for Silk (or the desktop
the user is using) to start them so it can make the api available to the
app and have control over the process so it can stop the app when the user
is done with it. Though if I use VoidSpace I could use that instead for the
api.

Take in account that servers need to finish someway, so if they are not
stateless AND designed to die themselves like oneshoot is, then you'll need
to end the backend process someway to don't left it running and you'll also
need to have some mechanism to recover if the backend server dies... I
think when this gets more mature I'll made a library based on oneshoot to
make this process simpler.

**Comment by piranna** [comment]
> One concern I have is Silk will have one way of running when in NodeOS
and another when running on any other operating system.

Why you think so?

**Comment by zodern** [comment]
> Why you think so?

I do not understand oneshoot enough. I am going to spend this weekend working on Silk and let you know then.

**Comment by lite20** [comment]
> @lite20 and @zodern, what do you think about this scheme to start backend servers for the GUI apps?

I like this implementation, It's simple. ^_^

I do have  a few questions though. Could services be remotely started with this? That could become a security issue as once a malicious script is on the device, a hacker could easily execute it via this interface. Are there plans to add some sort of authentication?

> > When accessing icon.png from the steam-ui server after it has been
started, would I use http://12.34.56.78/[/Steam/steam-ui]/hl3/icon.png or
http://12.34.56.78/hl3/icon.png or something else ?

> Good quedtion, since this was the use case that made me bang the head
against the wall all the weekend. It should be used relative paths, it's
said, hl3/icon.png, this way it gets transparent to the brackets section in
the url both if using its own web server or by using the static one from
oneshoot.

If putting the application in brackets would launch the application, wouldn't this request launch a new instance of the application? Infinitely calling requests and no end would actually be reached?

How would we go about implementing it? I don't know enough about websockets in nodejs but in the example you bound to a path. I didn't even know that was do-able :stuck_out_tongue: After you implement everything, a bit of experimentation should suffice to get a working system off it.

My understanding of oneshoot is also still extremely cloudy. If you could write an example implementation or even a super basic tutorial explaining what's happening behind the scenes that would be incredible :smile: 

The codefest continues...

**Comment by piranna** [comment]
> Could services be remotely started with this?

More than services, GUI apps backend server. But yes, that's the idea.

> That could become a security issue as once a malicious script is on the
device, a hacker could easily execute it via this interface. Are there
plans to add some sort of authentication?

Yes, all requests are authenticated using HTTP Auth agains the user name
and password used by logon :-) Currently this is done using HTTP Basic Auth
that send passwords in clear text so someone sniffing could get them, but
combined with SSL this is more secure and faster than Digest Auth, so I'm
thinking we could made HTTPS mandatory, the problem is the management of
certificates, since this would be common for all users... :-/ Maybe the
administrator boot mode could do the trick to install them :-)

> If putting the application in brackets would launch the application,
wouldn't this request launch a new instance of the application? Infinitely
calling requests and no end would actually be reached?

No, it isn't :-) Servers are started by nodeos-reverse-proxy, and it has a
map from the server path to the server instance, so if there's an entry for
that path, it doesn't start a new server instance :-)

> How would we go about implementing it? I don't know enough about
websockets in nodejs but in the example you bound to a path. I didn't even
know that was do-able

Magic, you know ;-) When I found that trick was possible some years ago I
was as surprised as you :-D

> My understanding of oneshoot is also still extremely cloudy. If you could
write an example implementation or even a super basic tutorial explaining
what's happening behind the scenes that would be incredible

I'm going to rewrite oneshoot as a library. Think of it as a normal server
created with the Node.js 'http' module that close itself when there are no
more open connections, there's no more than that. The other half is done by
nodeos-reverse-proxy, that start the oneshoot server when a new connection
is incoming and there's no currently a (oneshoot) server to listen to it,
no more. Hope that helps to understand how they work together :-)

**Comment by zodern** [comment]
> Think of it as a normal server
created with the Node.js 'http' module that close itself when there are no
more open connections, there's no more than that. The other half is done by
nodeos-reverse-proxy, that start the oneshoot server when a new connection
is incoming and there's no currently a (oneshoot) server to listen to it,
no more. Hope that helps to understand how they work together :-)

That's what I needed! Thanks. My concern is gone.

**Comment by piranna** [comment]
> That's what I needed! Thanks. My concern is gone.

Cool dude ;-)

**Comment by zodern** [comment]
> Regarding files access, the simple and obvious thing is use directly HTTP methods.

Sorry I didn't explain what I meant well. I was referring to clicking a file in file explorer and then having another app open the file. Silk has a property in the [app.json](https://github.com/Silk-GUI/text-editor/blob/master/app.json#L7) for what files the app can open and a [client api](https://github.com/Silk-GUI/text-editor/blob/master/public/js/textEditor.js#L5) for when it is supposed to open a file. Currently Silk opens a window for the app that will open the file and adds a query string with the file path. There probably is a better way of doing it.

> I think it's better to re-use package.json adding the custom fields that we would need (icon? anything more?) instead of add a new manifest file...

I agree. There is the icon, window url, if it supports multiple windows, if it needs any other ports available remotely, types of files it opens, and the app name that all would need custom fields. I can use a bower.json for the bower dependencies and get the window title from the html.

> That's strange, the 5 seconds only take effect when there's no more open connections to that user server (both HTTP and WebSocket) and get cancelled and renewed when a new one occurs... I've just reviewed the code and didn't find any problem here... :-/

Silk gives apps 5 seconds to start, and qemu without kvm was too slow. It wasn't because of oneshoot.

 > Apps can be "stored" inside npm packages, augmenting the package.json file with some fields like icon and others, and pure web apps could set the main field to its index.html. Hybrid web apps would be regular Node.js apps with its own webserver and so, being that at server.js (the default for npm start).

I was thinking about this last spring! I was also trying to figure out how to use npm for downloading, updating and removing apps but I am not sure I want apps to be inside `node_modules`.

**Comment by piranna** [comment]
> I was referring to clicking a file in file explorer and then having another app open the file.

This is a totally different discussion. I think this is better handled by using [protocol handlers](https://developers.google.com/web/updates/2011/06/Registering-a-custom-protocol-handler) and [content handlers](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/registerContentHandler).


> I can use a bower.json for the bower dependencies and get the window title from the html.

Ok for the window title thing. Why need to use bower? We can get the dependencies from npm too if the browser libraries are included as it's done in the main ones...


> Silk gives apps 5 seconds to start, and qemu without kvm was too slow. It wasn't because of oneshoot.

Ah, ok. Maybe is there any other way to check that they start that's not using a timeout?


> I am not sure I want apps to be inside node_modules.

Obviously this is not mandatory, and my brackets scheme allow to have apps in any place, but what problems would we have by having the apps inside `node_modules`? I think it's its natural place, specially if they are managed by npm... This way they can have later an entry on `/bin`, too...

**Comment by piranna** [comment]
I've just updated `onshoot` to make it `require()`able, but didn't test it yet. Take a look on it to see if you understand it better ;-) I've also changed it to allow to send the port over the communication channel if executed using `child_process.fork()` :-)

**Comment by lite20** [comment]
> Think of it as a normal server
created with the Node.js 'http' module that close itself when there are no
more open connections, there's no more than that. The other half is done by
nodeos-reverse-proxy, that start the oneshoot server when a new connection
is incoming and there's no currently a (oneshoot) server to listen to it,
no more. Hope that helps to understand how they work together :-)

It all makes sense now. XD

**Comment by piranna** [comment]
I've found that the latests changes on [Cloud9](https://c9.io) almost implement *exactly* what I was thinking for a GUI, but not also that. it's [free software](https://github.com/c9/core), it's published on [npm](https://www.npmjs.com/package/c9) and follow [the same scheme](https://github.com/c9/core/blob/master/server.js) of server + URL I was thinking about! Yay! :-D

If not as GUI desktop, at least we need to make this work as IDE :-)

**Comment by lite20** [comment]
> I've found that the latests changes on Cloud9 almost implement exactly what I was thinking for a GUI, but not also that. it's free software, it's published on npm and follow the same scheme of server + URL I was thinking about! Yay! :-D If not as GUI desktop, at least we need to make this work as IDE :-)

I'm not quite sure what you're referring to. Could you suace over some links? Has Cloud9 released something we should be looking into?

**Comment by lite20** [comment]
Hahaha leaving this thread open for discussion. If it stays inactive, we can close it
